crates/tarball/src/lib.rs

```diff
@@ -164,33 +164,46 @@ pub async fn download_tarball_to_store(
             integrity: package_integrity.to_string(),
             error,
         })?;
-    let url = package_url.to_string(); // TODO: use Arc
-    let cas_paths = tokio::task::spawn_blocking(move || {
-        verify_checksum(&response, package_integrity)
-            .map_err(|error| VerifyChecksumError { url, error })?;
-        let data = decompress_gzip(&response, package_unpacked_size)?;
-        Archive::new(Cursor::new(data))
-            .entries()?
-            .filter(|entry| !entry.as_ref().unwrap().header().entry_type().is_dir())
-            .map(|entry| -> Result<(OsString, PathBuf), TarballError> {
-                let mut entry = entry.unwrap();
-
-                // Read the contents of the entry
-                let mut buffer = Vec::with_capacity(entry.size() as usize);
-                entry.read_to_end(&mut buffer).unwrap();
-
-                let entry_path = entry.path().unwrap();
-                let cleaned_entry_path =
-                    entry_path.components().skip(1).collect::<PathBuf>().into_os_string();
-                let integrity = pacquet_cafs::write_sync(&store_dir, &buffer)?;
-
-                Ok((cleaned_entry_path, store_dir.join(integrity)))
-            })
-            .collect::<Result<HashMap<OsString, PathBuf>, TarballError>>()
-    })
-    .await
-    .expect("no join error")?
-    .pipe(Arc::new);
+    enum TaskError {
+        Checksum(ssri::Error),
+        Other(TarballError),
+    }
+    let cas_paths =
+        tokio::task::spawn_blocking(move || -> Result<HashMap<OsString, PathBuf>, TaskError> {
+            verify_checksum(&response, package_integrity).map_err(TaskError::Checksum)?;
+            let data =
+                decompress_gzip(&response, package_unpacked_size).map_err(TaskError::Other)?;
+            Archive::new(Cursor::new(data))
+                .entries()
+                .map_err(TarballError::Io)
+                .map_err(TaskError::Other)?
+                .filter(|entry| !entry.as_ref().unwrap().header().entry_type().is_dir())
+                .map(|entry| -> Result<(OsString, PathBuf), TarballError> {
```

Comment:
Out of context for this pull request, but why don't we use `filtermap` in here?

Reply:
> why don't we use `filtermap` in here?

The author of the PR that added `filter` didn't put it there.

I will try `filter_map` in monday. But if it makes the code more complex than necessary, I will revert it.

Reply:
Actually, `filter_map` requires the function inside to return an `Option`. It would make the signature of the function more complicated. I also don't think where should zealously replace all instances of `.filter.map` with a `filter_map`. `filter_map` is just a more specific version of `flat_map`, which is meant to replace `.map.flatten`.